### ISSUE TITLE
 public async Task<TimedEntityTagHeaderValue> QueryAsync(HttpContext context) is not executed

### ISSUE THREAD

**Issue by ahedreville** [aliostad/CacheCow#224]
Hi I had a working code that suddenly is not working any more. After debugin the code I see that all works fine, ITimedETagQueryProvider and ITimedETagExtractor are instanciated with http scoped and runs, but for the ITimedETagQueryProvider the QueryAsync is never executed. I am wondering why, since was working on past and no changes has been done on the code. See the relevant net core code here.

Controler code
```
        [HttpCacheFactory(0, ViewModelType = typeof(DepartmentDto))]
        public async Task<IActionResult> GetDepartments(....
```

Registration
```
            services.AddHttpCachingMvc();

            services.AddQueryProviderForViewModelMvc<DepartmentDto, RowVersionETagQueryProvider<Department, DepartmentDto>>();
            services.AddScoped<ITimedETagExtractor<DepartmentDto>, RowVersionETagExtractor<Department, DepartmentDto>>();       // ITimedETagExtractor must be resistered after TQueryProvider registration.

```
TagQueryProvider It works as expected no proble with this code
BUT THE QueryAsync IS NEVER EXECUTED
```
public class RowVersionETagQueryProvider<TEntity, TEntityDto> : ITimedETagQueryProvider<IEnumerable<TEntityDto>>, ITimedETagQueryProvider<TEntityDto> where TEntity : class , IRowVersion where TEntityDto : class , IRowVersion 
    {
        private readonly bool _dispose;
        private readonly IUserFunctionsContext _context;

        public RowVersionETagQueryProvider(UserFunctionsContext context)
        : this(context, false)
        {
        }

        public RowVersionETagQueryProvider(IUserFunctionsContext context, bool dispose)
        {
            _context = context;
            _dispose = dispose;
        }

        public void Dispose()
        {
            Dispose(_dispose);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
                _context.Dispose();
        }

        ~RowVersionETagQueryProvider() => Dispose(_dispose);

        public async Task<TimedEntityTagHeaderValue> QueryAsync(HttpContext context)
        {
            byte[] rowVer;

            var routeData = context.GetRouteData();

            var queryString = context.Request.QueryString;

            
            if (routeData.Values.ContainsKey(nameof(IEntity<int>.Id)))
            {
                rowVer = await _context.GetRowVersionAsync<TEntity>(routeData.Values[nameof(IEntity<int>.Id)]).ConfigureAwait(false);
            }
            else
            {
                rowVer = await _context.GetRowVersionAsync<TEntity>().ConfigureAwait(false);
            }

            var queryBytes = Encoding.UTF8.GetBytes(queryString.HasValue ? queryString.Value : "");

            if (rowVer is byte[] maxRowVer && maxRowVer.Length > 0)
            {
                if (queryBytes != null && queryBytes.Length > 0)
                    rowVer = maxRowVer.Concat(queryBytes).ToArray();
            }
            else
            {
                if (queryBytes != null && queryBytes.Length > 0)
                    rowVer = queryBytes;
            }

            return new TimedEntityTagHeaderValue(rowVer.ToETagString());
        }

   }
```

The  RowVersionETagExtractor.. It works as expected no proble with this code

```
public class RowVersionETagExtractor<TEntity, TEntityDto> : ITimedETagExtractor<TEntityDto>, IDisposable where TEntity : class, IRowVersion where TEntityDto : class, IRowVersion, new()
    {
        private readonly bool _dispose;
        private readonly IUrlHelper _urlHelper;
        private readonly IUserFunctionsContext _context;

        public RowVersionETagExtractor(UserFunctionsContext context, IUrlHelper urlHelper)
        : this(context, urlHelper, false)
        {
        }

        public RowVersionETagExtractor(IUserFunctionsContext context, IUrlHelper urlHelper, bool dispose)
        {
            _context = context;
            _dispose = dispose;
            _urlHelper = urlHelper;
        }

        public void Dispose()
        {
            Dispose(_dispose);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
                _context.Dispose();
        }

        ~RowVersionETagExtractor() => Dispose(_dispose);

        public TimedEntityTagHeaderValue Extract(TEntityDto dto) => new TimedEntityTagHeaderValue(dto.RowVersion.ToETagString());

        public TimedEntityTagHeaderValue Extract(object viewModel)                           // view model is an object box that could contain five different types of objects.
        {
            byte[] rowVer = null;
            var mockedDto = new TEntityDto();

            var queryString = _urlHelper
                .ActionContext
                .HttpContext
                .Request
                .QueryString;

            switch (viewModel)
            {
                case TEntityDto dto:                                                         // (1) view model dto don't hasn't been shaped. View model is a regular dto class.
                    rowVer = dto.RowVersion;
                    break;

                case IEnumerable<TEntityDto> _:                                              //(2) view model enumerable dto hasn't been shaped. View model is a regular enumerable collection of dto classes.
                    rowVer = Task.Run(async () =>
                        await _context.GetRowVersionAsync<TEntity>().ConfigureAwait(false)).Result;
                    break;

                case IDictionary<string, object> shapedDto:                                  // (3) view model dto has been shaped. View model is an expando object built from fields parameters over a regular dto class.
                    shapedDto.TryGetValue(nameof(IRowVersion.RowVersion), out var value);
                    rowVer = value as byte[];
                    break;

                case IEnumerable<IDictionary<string, object>> _:                             // (4) view model enumerable dto has been shaped. View model is an enumerable collection of expando object built from fields parameters over a regular enumerable collection of dto classes.
                    rowVer = Task.Run(async () =>
                        await _context.GetRowVersionAsync<TEntity>().ConfigureAwait(false)).Result;
                    break;

                case LinkedDtosWrapper _:                                                    // (5) view model enumerable linked dto collection has been wrapped. View model wrapper with a values property with an enumerable collection of expando object built from fields parameters over a regular enumerable collection of dto classes.
                    rowVer = Task.Run(async () =>
                        await _context.GetRowVersionAsync<TEntity>().ConfigureAwait(false)).Result;
                    break;
            }

            var requestQueryBytes = Encoding.UTF8.GetBytes(queryString.HasValue ? queryString.Value : "");

            if (rowVer != null && rowVer.Length > 0)
            {
                if (requestQueryBytes != null && requestQueryBytes.Length > 0)
                {
                    mockedDto.RowVersion = rowVer.Concat(requestQueryBytes).ToArray();
                }
                else
                {
                    mockedDto.RowVersion = rowVer;
                }
            }
            else
            {
                if (requestQueryBytes != null && requestQueryBytes.Length > 0)
                    mockedDto.RowVersion = requestQueryBytes;
            }

            return Extract(mockedDto);
        }
    }

```

**Comment by ahedreville** [comment]
I am reducing the minimal the code and same problem..

Model and View definition..

```
 public class Foo
    {
    }

    public class FooView
    {
    }
```
DI registration

```
            services.AddQueryProviderForViewModelMvc<FooView, RowVersionETagQueryProvider2<Foo, FooView>>();
            services.AddScoped<ITimedETagExtractor<FooView>, RowVersionETagExtractor2<Foo, FooView>>();       // ITimedETagExtractor must be resistered after TQueryProvider registration.

```

Controller

```
        [HttpCacheFactory(0, ViewModelType = typeof(FooView))]
        public async Task<IActionResult> GetDepartments(...

```
Mininal ETagQueryProvider

```
public class RowVersionETagQueryProvider2<TEntity, TEntityDto> : ITimedETagQueryProvider<IEnumerable<TEntityDto>>, ITimedETagQueryProvider<TEntityDto> where TEntity : class where TEntityDto : class
    {

        public void Dispose()
        {
        }

        public async Task<TimedEntityTagHeaderValue> QueryAsync(HttpContext context)
        {
            await Task.CompletedTask.ConfigureAwait(false);

            return new TimedEntityTagHeaderValue("FOO");
        }

   }

```
Minimal ETagExtractor

```
    public class RowVersionETagExtractor2<TEntity, TEntityDto> : ITimedETagExtractor<TEntityDto>, IDisposable where TEntity : class where TEntityDto : class
    {

public TimedEntityTagHeaderValue Extract(TEntityDto dto) => new TimedEntityTagHeaderValue("FOO");

        public TimedEntityTagHeaderValue Extract(object viewModel)
        {

            return Extract(null);
        }

        public void Dispose()
        {
        }
    }
```

Same issue!!
ETagQueryProvider and ETagExtractor are scoped instantiated when postmant hits the controler and disposed..  Extract method is executed BUT QueryAsync gets NEVER EXECUTED ;(

**Comment by ahedreville** [comment]
It seems that I would need to refactor the code, to follow the registrations showed on the samples

```
            services.AddHttpCachingMvc();
            services.AddQueryProviderAndExtractorForViewModelMvc<Car, TimedETagQueryCarRepository, CarETagExtractor>(false);
            services.AddQueryProviderAndExtractorForViewModelMvc<IEnumerable<Car>, TimedETagQueryCarRepository, CarCollectionETagExtractor>(false);
```

I will do it return back... my inputs..

**Comment by ahedreville** [comment]
Yes I see my problem..
It is because I am doing a "tricky" registration on the dependency injection container.
The reason is that really I need to have two generic arguments for my ETagExtractor.

Let me explain a bit. starting from a fresh new example..

Here the startup.. code..

```
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_1);

            services.AddScoped<DataContext>();

            services.AddHttpCachingMvc();

            services.AddQueryProviderAndExtractorForViewModelMvc<Department, RowVersionETagQueryProvider<Department>, RowVersionETagExtractor<Department>>(true);

            //services.AddQueryProviderAndExtractorForViewModelMvc<IEnumerable<Department>, RowVersionETagQueryProvider, RowVersionCollectionETagExtractor>(false);
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, DataContext context)
        {
            context.Database.EnsureCreated();

            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseMvc();
        }
    }

```
 and a InMemory dbcontext...

```
   public class DataContext : DbContext
    {
        public DbSet<Department> Department { get; set; }
        
        protected override void OnModelCreating(ModelBuilder builder)
        {
            builder.Entity<Department>(e =>
            {
                e.HasData(
                    new object[] {
                        new {Id = 01, Name = "Engineering"},
                        new {Id = 02, Name = "Tool Design"},
                        new {Id = 03, Name = "Sales"},
                        new {Id = 04, Name = "Marketing"},
                        new {Id = 05, Name = "Purchasing"},
                        new {Id = 06, Name = "Research and Development"},
                        new {Id = 07, Name = "Production"},
                        new {Id = 08, Name = "Production Control"},
                        new {Id = 09, Name = "Human Resources"},
                        new {Id = 10, Name = "Finance"},
                        new {Id = 11, Name = "Information Services"},
                        new {Id = 12, Name = "Document Control"},
                        new {Id = 13, Name = "Quality Assurance"},
                        new {Id = 14, Name = "Facilities and Maintenance"},
                        new {Id = 15, Name = "Shipping and Receiving"},
                        new {Id = 16, Name = "Executive"}
                    });
            });
        }

        protected override void OnConfiguring(DbContextOptionsBuilder builder)
        {
            base.OnConfiguring(builder);

            if (!builder.IsConfigured)
                builder.UseInMemoryDatabase("InMemory");
        }
    }

```
Ok.. lets continue...

CacheCow needs a TView type for the HttpCacheFactoryAttribute as follows..

`[HttpCacheFactory(0, ViewModelType = typeof(Department))]`

And what I am using is a domain model like...

```
    public class Department
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }
```

Taking a look to the controller

    [Route("api/[controller]")]
    [ApiController]
    public class ValuesController : ControllerBase
    {
        private readonly DataContext _context;

        public ValuesController(DataContext context)
        {
            _context = context;
        }

```
        [HttpGet]
        [HttpCacheFactory(0, ViewModelType = typeof(Department))]
        public ActionResult<IEnumerable<string>> Get()
        {
            var dtos = _context.Department.Select(e => new DepartmentDto {Id = e.Id, Name = e.Name, Age = DateTimeOffset.Now}).ToArray();

            return Ok(dtos);
        }

        [HttpGet("{id}")]
        [HttpCacheFactory(0, ViewModelType = typeof(Department))]
        public ActionResult<string> Get(int id)
        {
            var e = _context.Department.Find(id);
            var dto  = new DepartmentDto {Id = e.Id, Name = e.Name, Age = DateTimeOffset.Now};

            return Ok(dto);
        }

    }
```
What I am returning is not a domain object like Department.. I am returning a data transfer object

```
    public class DepartmentDto
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public DateTimeOffset Age { get; set; }
    }
```

But for my  RowVersionETagExtractor and RowVersionETagQueryProvider is crucial to know the model domain type in order to use and entity framework dbset..

So for the RowVersionETagExtractor having TModel as I have for the controller  attribute [HttpCacheFactory(0, ViewModelType = typeof(Department))] will suffice and  all are happy the since the  ITimedETagQueryProvider<TModel> has the same type that the controller attribute, and I am able to get the _context.Set<TModel>() inside my QueryAsync that this time YES IT GETS EXECUTED..

```
    public class RowVersionETagQueryProvider<TModel> : ITimedETagQueryProvider<TModel> where TModel : class
    {
        private readonly DataContext _context;

        public RowVersionETagQueryProvider(DataContext context)
        {
            _context = context;
        }

        public void Dispose()
        {
        }

        public Task<TimedEntityTagHeaderValue> QueryAsync(HttpContext context)
        {

            var foo = _context.Set<TModel>().FirstOrDefault(); //Do Whatever do you need with your repository

            return Task.FromResult(new TimedEntityTagHeaderValue("FOO"));
        }
    }
```

But the

**Comment by ahedreville** [comment]
Hi again

refactoring the registration and the ETagExtractor as follows finally will work as I need..

`services.AddQueryProviderAndExtractorForViewModelMvc<Department, ETagQueryProvider<Department>, ETagExtractor<Department, DepartmentDto>>();`


```
    public class ETagExtractor<TModel, TView> : ITimedETagExtractor<TModel> where TModel : class where TView : class
    {
        private readonly DataContext _context;

        public ETagExtractor(DataContext context)
        {
            _context = context;
        }

        public TimedEntityTagHeaderValue Extract(object viewModel)
        {
            var foo = _context.Set<TModel>().FirstOrDefault();

            return new TimedEntityTagHeaderValue("FOO");
        }

        public TimedEntityTagHeaderValue Extract(TModel viewModel)
        {                                                            
            return Extract(viewModel as TView);
        }
    }

```
but remember that the code branch `public TimedEntityTagHeaderValue Extract(TModel viewModel)` will be **never executed.**

However the `public TimedEntityTagHeaderValue Extract(object viewModel)` **will be executed** and you already knows that **viewModel should be an object of type TView**.

Happy.. Yes!!! ;-)

**Comment by ahedreville** [comment]
I am getting mad..
now I see that changing the launchSettings.json will reproduce the issue.. I will need to look why..

**Comment by ahedreville** [comment]
Ok.. I will continue the discussion..
First of all just remind me that ll I had said about having a second second generic parameter and use the extension method for registration and that one of the branch of code of the ETagExtractor will not be executed remains true.

Buth when I "refactored my conclusions" from the tiny demo ware to my global application I faced the same problem again, I mean **QueryAsync IS NEVER EXECUTED**.

So I did a more deeply investigation on the demo ware again.. and  I will paste some inputs and images

![snap281](https://user-images.githubusercontent.com/7416613/47253487-d2465980-d453-11e8-9fa2-bfe7764f3f7c.jpg)
![snap283](https://user-images.githubusercontent.com/7416613/47253489-dbcfc180-d453-11e8-9400-3c01dc42ab2d.jpg)

**Comment by ahedreville** [comment]
Can you see the problem ?..
When I use the chrome browser everything runs fine.. but when I use postman, despite of being instantiated the ETagQueryProvider, the **QueryAsync IS NEVER EXECUTED**.

Why?.. really I don't know..

Just highlight that on my global app I am using Kestrel not IIS and postman for testing, so lets change the launchSettings.json from this..

```
{
  "$schema": "http://json.schemastore.org/launchsettings.json",
  "iisSettings": {
    "windowsAuthentication": false, 
    "anonymousAuthentication": true, 
    "iisExpress": {
      "applicationUrl": "http://localhost:55877",
      "sslPort": 44300
    }
  },
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "launchUrl": "api/values",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "WebApplication1": {
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "api/values",
      "applicationUrl": "https://localhost:5001;http://localhost:5000",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

to this..

```
{
  "profiles": {
    "WebApplication1": {
      "commandName": "Project"
    }
  }
}
```
And let see what we have now...
and paste a development time  localhost.pfx certificate file on the root folder.

Let change the appsettings.json to use Kestrel as follows..

```
{
    "Logging": {
        "LogLevel": {
            "Default": "Warning"
        }
    },
    "AllowedHosts": "*",

    "Kestrel": {
        "Certificates": {
            "Default": {
                "Path": "localhost.pfx",
                "Password": "3219b7490e1a47beb4a222869390ec68"
            }
        },
        "EndPoints": {
            "Https": {
                "Url": "https://localhost:44310"
            }
        }
    }
}

```

**Comment by ahedreville** [comment]
![snap284](https://user-images.githubusercontent.com/7416613/47253827-3881ab00-d459-11e8-846f-619407b2044f.jpg)

Ok I have the same scenario...

**When I use the chrome browser everything runs fine.. but when I use postman**, despite of being instantiated the ETagQueryProvider, the **QueryAsync IS NEVER EXECUTED**.

Why?

I would need some help to resolve this...

Is a perhaps a BUG?..

Perhaps the CacheCow.Server is failing to get postman headers and see that I am sending the If-None-Match header..?

Lets see what we have on the controler action..

I add an IUrlHelper registration to get access to the context where the action is executed as follows

```
            services
                .AddSingleton<IActionContextAccessor, ActionContextAccessor>()                                   // Action context accessor must be registered as singleton and before registering the UrlHelper as scoped.
                .AddScoped<IUrlHelper>(s=> new UrlHelper(s.GetService<IActionContextAccessor>().ActionContext)); // UrlHelper must be registered as scoped ant it will need the context accessor the to get the context in which the action runs.

```

And now I refactor the controller as follows..

```
    [Route("api/[controller]")]
    [ApiController]
    public class ValuesController : ControllerBase
    {

        protected readonly IUrlHelper _urlHelper;
        private readonly DataContext _context;

        public ValuesController(DataContext context, IUrlHelper urlHelper)
        {
            _urlHelper = urlHelper;
            _context = context;
        }

        [HttpGet]
        [HttpCacheFactory(0, ViewModelType = typeof(Department))]
        public ActionResult<IEnumerable<string>> Get()
        {
            var headerkeys = _urlHelper.ActionContext.HttpContext.Request.Headers.Keys.ToArray();
            var headervalues = _urlHelper.ActionContext.HttpContext.Request.Headers.Values.ToArray();

            var dtos = _context.Department.Select(e => new DepartmentDto {Id = e.Id, Name = e.Name, Age = DateTimeOffset.Now}).ToArray();

            return Ok(dtos);
        }

        [HttpGet("{id}")]
        [HttpCacheFactory(0, ViewModelType = typeof(Department))]
        public ActionResult<string> Get(int id)
        {
            var e = _context.Department.Find(id);
            var dto  = new DepartmentDto {Id = e.Id, Name = e.Name, Age = DateTimeOffset.Now};

            return Ok(dto);
        }

    }
```

I set a break point on the controller action, started the application and fired postmand, and let see what we have as request headers..

**Comment by ahedreville** [comment]
aha! **something has happened to my If-None-Match tag when using postman**, I added some extra headers and **the only missed on the controller action context is the If-None-Match** header.

![snap285](https://user-images.githubusercontent.com/7416613/47254136-f5760680-d45d-11e8-8f4a-7563d38a34d6.jpg)


Weird..

**Comment by ahedreville** [comment]
Ok.. **I will try now insomnia** another rest client

![snap286](https://user-images.githubusercontent.com/7416613/47254843-8dc4b900-d467-11e8-8019-f9eea0927c94.jpg)

And aha!! **the If-None-Match tag is present**...

**Thus CacheCow.Server as expected**. **The ETagQueryProvider method QueryAsync IS EXECUTED**.

The problem is what is doing postman **something is steeling my If-None-Match header from postman**.

**Comment by ahedreville** [comment]
**I can assure that CacheCow  is working perfect!!!**

The problem is on the postman rest client.

I downloaded an old portable version 6.0.9 of postman and  CacheCow worked as expected..

However my current installed version of postman 6.4.4 was steeling my If-None-Match header.

I had removed the current installed version and deleted the ..\AppData\Local\postman and ..\AppData\Roaming\postman folder from my profile and downloaded and installed the last version of postman this is the same 6.4.4 that I was using..

**et voila!! postman is sending back again the If-None-Match header**

**So indeed CacheCow is working and the reason why the ETagQueryProvider method QueryAsync wasn't executed is due a corrupted installation of postman that was not sending the If-None-Match header**.

Thanks to this debugging session I had learned a lot more about CacheCow 

Thanks to all... **This issue wasn't a CacheCow issue**..

So this thread could be closed now.
